### PR TITLE
Check append

### PR DESCRIPTION
**Pull request by Rlamboll**
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Example added (either to an existing notebook or as a new notebook, where applicable)
- [x] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <https://github.com/znicholls/silicone/pull/XX>`_) Added feature which does something
- (`#XX <https://github.com/znicholls/silicone/pull/XX>`_) Fixed bug identified in (`#YY <https://github.com/znicholls/silicone/issues/YY>`_)
```

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -4,6 +4,7 @@ Changelog
 master
 ------
 
+- (`#83 <https://github.com/znicholls/silicone/pull/83>`_) Ensured that all results of crunching can be appended to the input.
```
Comment (znicholls) [inline CHANGELOG.rst:7]:
```suggestion
- (`#83 <https://github.com/znicholls/silicone/pull/83>`_) Add test that all results of crunching can be appended to the input.
```

File: tests/integration/crunchers/test_cruncher_constant_ratio.py [hunk 2]
```diff
@@ -117,6 +120,9 @@ def test_relationship_usage(self, unequal_df, test_downscale_df):
             test_downscale_df.timeseries().columns.values.squeeze(),
         )
 
+        # Test we can append the results correctly
+        test_downscale_df.append(res)
```
Comment (znicholls) [inline tests/integration/crunchers/test_cruncher_constant_ratio.py:124]: I would suggest making this test slightly more robust by checking that the data actually stayed in the append dataframe and didn't explode. That could be done by adding something like

```python
res_from_append = test_downscale_df.filter(whatever_kwargs_required_to_get_res_back_out)

# can't remember what the pyam testing function is but hope the idea makes sense...
pyam.testing.compare_iamdfs(res, res_from_append)

# you could also test that extra columns went in as expected e.g.
# might require some indexing here as you might want different values at different points depending on how 
# the auto-infilling is meant to work
test_downscale_df[extra_col] == expected_result  
```

File: tests/integration/crunchers/test_cruncher_interpolate_specified.py [hunk 3]
```diff
@@ -171,6 +176,9 @@ def test_relationship_usage(self, simple_df):
         assert all(expect_01.filter(year=2030)["value"] == 1000)
         assert all(expect_01.filter(year=2050)["value"] == 5000)
 
+        # Test we can append our answer
+        simple_df.filter(variable=lead).append(expect_01)
```
Comment (znicholls) [inline tests/integration/crunchers/test_cruncher_interpolate_specified.py:180]: as above re extra check

File: tests/integration/crunchers/test_cruncher_latest_time_ratio.py [hunk 4]
```diff
@@ -174,6 +177,9 @@ def test_relationship_usage(self, test_db, test_downscale_df):
             test_downscale_df.timeseries().columns.values.squeeze(),
         )
 
+        # Test that we can append the output to the input
+        test_downscale_df.filter(variable=lead).append(res)
```
Comment (znicholls) [inline tests/integration/crunchers/test_cruncher_latest_time_ratio.py:181]: as above re extra check

File: tests/integration/crunchers/test_cruncher_quantile_rolling_windows.py [hunk 5]
```diff
@@ -234,6 +239,9 @@ def test_extreme_values_relationship(self):
         extreme_crunched = res(modify_extreme_db)
         assert all(crunched["value"] == extreme_crunched["value"])
 
+        # Check that we can append the answer
+        large_db.filter(variable=lead).append(crunched)
```
Comment (znicholls) [inline tests/integration/crunchers/test_cruncher_quantile_rolling_windows.py:243]: as above re extra check

File: tests/integration/crunchers/test_cruncher_rms_closest.py [hunk 6]
```diff
@@ -370,6 +377,9 @@ def test_relationship_usage(self, test_db, test_downscale_df):
             exp.timeseries().columns.values.squeeze(),
         )
 
+        # Test we can append the answer to the original
+        test_downscale_df.filter(variable=lead).append(res)
```
Comment (znicholls) [inline tests/integration/crunchers/test_cruncher_rms_closest.py:381]: as above re extra check

File: tests/integration/crunchers/test_cruncher_time_dep_ratio.py [hunk 7]
```diff
@@ -155,6 +160,9 @@ def test_relationship_usage_multiple_data(
             test_downscale_df.timeseries().columns.values.squeeze(),
         )
 
+        # Test we can append our answer
+        test_downscale_df.filter(variable=lead).append(res)
```
Comment (znicholls) [inline tests/integration/crunchers/test_cruncher_time_dep_ratio.py:164]: as above re extra check

File: CHANGELOG.rst (hunk 1 above)
Comment (Rlamboll) [reply CHANGELOG.rst:7]: similar commit added

File: tests/integration/crunchers/test_cruncher_constant_ratio.py (hunk 2 above)
Comment (Rlamboll) [reply tests/integration/crunchers/test_cruncher_constant_ratio.py]: I've added the comparison between the filtered and the output here. I check for the column genuinely being added to the original directly after it is added.

Comment (znicholls) [reply tests/integration/crunchers/test_cruncher_constant_ratio.py]: I meant more like https://github.com/znicholls/silicone/pull/84, just to make sure the appending hasn't exploded (there's also the question of what should happen if you have more than one value in the extra column)

File: tests/integration/crunchers/test_cruncher_rms_closest.py [hunk 8]
```diff
@@ -370,6 +380,14 @@ def test_relationship_usage(self, test_db, test_downscale_df):
             exp.timeseries().columns.values.squeeze(),
         )
 
+        # Test we can append the answer to the original
+        appended_df = test_downscale_df.filter(variable=lead).append(res)
+        assert appended_df.filter(variable=follow).equals(res)
+        if add_col:
+            # Currently we infill with blanks because added columns do not correspond to
+            # unique entries
```
[the remark this answers is not in the document]
Comment (Rlamboll) [reply tests/integration/crunchers/test_cruncher_rms_closest.py:388]: No, good spot!

File: tests/integration/crunchers/test_cruncher_time_dep_ratio.py [hunk 9]
```diff
@@ -155,6 +163,10 @@ def test_relationship_usage_multiple_data(
             test_downscale_df.timeseries().columns.values.squeeze(),
         )
 
+        # Test we can append our answer
+        appended_df = test_downscale_df.filter(variable=lead).append(res)
+        assert appended_df.filter(variable=follow).equals(res)
```
[the remark this answers is not in the document]
Comment (Rlamboll) [reply tests/integration/crunchers/test_cruncher_time_dep_ratio.py:168]: Yeah, check added

**Comment by Rlamboll** [comment]
I have raised separate point #85 for additional columns ruining everything.

**Pull request merged**